pkg/controller/control.go

```diff
@@ -759,6 +765,31 @@ func (c *Controller) WriteAt(b []byte, off int64) (int, error) {
 	return n, err
 }
 
+func (c *Controller) writeInWOMode(b []byte, off int64) (int, error) {
+	sectorSize := int64(4096)
+	bufLen := int64(len(b))
+	// buffer b is sectorSize aligned
+	if (bufLen >= sectorSize) && (off%sectorSize == 0) && ((off+bufLen)%sectorSize == 0) {
+		return c.backend.WriteAt(b, off)
+	}
+
+	readOffsetStart := (off / sectorSize) * sectorSize
+	readOffsetEnd := (((off + bufLen) / sectorSize) + 1) * sectorSize
+	readBuf := make([]byte, readOffsetEnd-readOffsetStart)
```

Comment:
Think if it's possible to reuse buffer here instead of reallocation.

Reply:
Each read will potentially require a different readBuff length. I am not sure how to reuse this one

Reply:
This function needs to read the unaligned part (of which the length is smaller than the `defaultSectorSize`) only, hence a buffer with a fixed length `defaultSectorSize` should be enough.

Reply:
1. Since this function `writeInWOMode` is called concurrently by multiple go routines, I think it is not safe to pre-allocated and reuse the buffer
2. Suppose that the incoming IO is from offset 4000 to 9000, if we use fixed length 4096 for the readBuff, we would have to do 3 read IOs (0-4095, 4096-8191, 8192-12287) then followed by 3 write IOs. These read IO potentially will have to go over the network. The benefit is we save a very tiny RAM. Therefore, I think we shouldn't use fixed 4096 length for the readBuff. What do you think?

Reply:
ping @innobead and @shuo-wu

Reply:
Agreed 👍